File: TrickyTropolis/Assets/Code/LevelGroup.cs
using UnityEngine;
using System.Collections;

public class Level
{
    public readonly string hand;
    public readonly int energyGoal;
	public readonly int bonusGoal;
	public string messageTitle;
	public string[] messages;
    public Level( string h, int e, int b = -1 )
    {
        hand = h;
        energyGoal = e;
		bonusGoal = b;
    }
	public Level addMessage( string title, params string[] m ) {
		messageTitle = title;
		messages = m;
		return this;
	}
}

public class LevelGroup {
    public readonly Level[] levels;
    public readonly string name;
	public readonly int trophiesSkippable;

    public LevelGroup(string n, Level[] lvls, int skippable=0 ) { 
		name = n; 
		levels = lvls;
		trophiesSkippable = skippable;
	}

    public static LevelGroup[] allGroups = {
	/*
	new LevelGroup ( "Test", new Level[] {
			new Level( "w w w", 2, 6 )
				.addMessage( "Title Bar", "First message", "second message", "third message" ),
			new Level("w iz iz cs cy s s s sh sh", 2, 46 ),
			new Level("w pg ps wt iz iz iz iz cs sh", 2, 50 )
	}),*/

	new LevelGroup ("Beginner", new Level[] {
        new Level ("w w w w", 8)
.addMessage("Welcome to TrickyTropolis!", 
"The object of this game is to reach the Energy goal for each level.  " +
"You gain Energy by creating buildings, each of which holds a certain amount of Energy.",
"To complete this level, simply play all four Warehouses by dragging them onto " + 
"empty slots.",
"Whenever you need to try a level again, press and hold the red button for 1 second to restart from the " +
"beginning.  There is no limit to the number of attempts you can make!"),
        new Level ("w w cs cs", 8)
.addMessage( "Energy Cost",
"Pay attention to the numbers in the upper right corner of your cards!  The first " +
"number is Energy Cost, and the second is Starting Energy.",
"You can take a closer look at your cards by tapping on them.  Tap the Charging Station now.",
"Because the Charging Station has a cost of 2, it cannot be played onto an empty slot.  " +
"Instead, it must be played over an existing building having 2 or more Energy.  That building " +
"and its energy will be lost, so choose wisely!"),				          
        new Level ("w w w w s", 9)
.addMessage( "Buildup",
"Some buildings have Buildup.  They will gain +1 Energy every time another new building is created.  " +
"These buildings can gain a lot of Energy if you play them early!"),
        new Level ("w w cs wt", 7)
.addMessage( "Supply",
"Some buildings have Supply.  They will provide +1 extra Energy to every new building you create.  " +
"Supply is another ability that you want to get as early as possible." ),			           
        new Level ("d w w wt", 6)
.addMessage( "Upgrade",
"A building with the Upgrade attribute will give a bonus to whichever building is constructed " +
"<i>over</i> it.  The Databank gives +1 extra energy.",
"You will need this extra Energy to construct a Wind Turbine."),
		new Level ("d w cs cs", 8)
.addMessage( "Energy Precision",
"In general, you can get the most Energy from your buildings when you meet their Energy Cost exactly.  " +
"For this level, make sure to waste nothing!"),					           
        new Level ("w iz iz s", 4)
.addMessage( "Industrial Zone", 
"The Industrial Zone creates two buildings with no starting Energy.  Although it does little by itself, " +
"it is very useful for triggering other abilities such as Buildup and Supply.",
"TrickyTropolis is all about making your cards work together!"),
        new Level ("d w wt ps", 7)
.addMessage( "Event Cards",
"Event cards are distinguished by their purple color.  Instead of creating a new building, they provide " +
"benefits to your existing buildings."),
		new Level ("d d cs rl", 6)
.addMessage( "Research Lab",
"The Research Lab is a special building with an Energy multiplier.  Whenever a building would " +
"normally gain energy, the Research Lab gains twice as much!",
"This makes it a perfect target for cards such as Energy Boost.")
    }, 0 ),

		new LevelGroup ("Student", new Level[] {
		new Level ("iz iz qb", 4)
			.addMessage( "Student", 
		"Welcome to the Student levels!  Now that you know the basics, " +
		"there will be fewer hints.  If you're not sure what a card does, experiment!  You can always " +
		"restart the level." ),
		new Level ("d d iz s s qb", 12),	
		new Level ("d w s s cy cy", 5)
			.addMessage( "Construction Yard", 
				"The Construction Yard lowers the cost of buildings in your hand by 1.  Not only does " +
				"it make your buildings cheaper to play, but also any building that normally costs 1 can be played " +
				"onto an empty slot!"),			
		new Level ("w s s iz cy qb", 7),
		new Level ("d w s cs rl ps", 12),
		new Level ("w w w dn dn", 10)
			.addMessage( "Distribution Netork",
		"The Distribution Network boosts the buildings adjacent to it when it is played.  Try to boost " +
		"as many buildings as you can!",
		"Buildings across the diagonal don't count as adjacent." ),

        new Level ("d iz dn", 4, 5)
			.addMessage( "Bonus Trophy", 
		"There is a bonus trophy in this level!  You can earn it if you fill your energy meter " +
		"<i>past</i> the top.  Trophies will help you unlock additional stuff in the game.",
		"We've told you about this one, but normally these bonus trophies will be hidden!  You " +
		"won't know which levels they're in until you find them!",
		"The Level Select screen will tell you how many trophies are still left to find within each " +
		"group." ),
		new Level ("w w cs sf rl ps", 13 )
			.addMessage( "Scaffolding",
			"The Scaffolding grants Buildup to the building played over it.  It works the " +
			"same way as the Databank.  You can give Buildup to potentially any building, if you have the " +
			"right cards!"),
		new Level("d w cs cs f", 9),
        new Level ("w w w sc cs rl", 10)
			.addMessage( "Storage Cell",
		"The Storage Cell starts with 1 energy, and also steals 1 more from each other building you have.",
		"This can give you a free building with a lot of starting energy, but be very careful about " +
		"which buildings to steal from!" )
	//new Level(
	//new Level(
    }, 0 ),

    new LevelGroup ("Challenger", new Level[] {
		new Level ("d w w s s s s s", 8)
			.addMessage( "Challenger", 
		"You made it!  Time to ramp up the difficulty even more!  You'll have more cards, and " +
		"there are two hidden trophies somewhere in these levels!" ),
		new Level ("w w w s cy wt ps", 11, 12), //Good candidate for trophy
		new Level ("iz w s cs qb qb", 11),
		new Level ("d d iz w sc wt", 10),
		new Level ("iz iz w s qb b", 11, 12 ), //Trophy?
		//new Level ( "d w s iz cs hd ps", 14 ),
		new Level ("d sc sc w dn rl", 10),	
		new Level("d d w s s sf cy f", 15),
		new Level ("d w w s wt wt wt wt", 9),
		new Level("w w d cs wt sf r", 15)
    }, 1 ),

	new LevelGroup ("Expert", new Level[] {
		new Level( "iz sc w w cs wt sk sk b", 28 ),
		new Level ("w w w dn dn cs cy f", 18, 19), //Trophy?
		new Level ("d d sc s cs cy rl", 9, 10), //Red Herring
		new Level ("d d w sf dn cs wt", 14, 15), //Hard!
		new Level( "d iz iz iz dn dn dn", 14, 15 ),
		new Level( "d d w sf cs wt rl ps", 20 ),  //Trophy?
		new Level( "w w w cs cs cs rl tt b ps", 27 ),
		new Level( "d w w iz cs wt wt wt cp ps", 25 ),
		new Level( "d d iz iz iz w cs cp ps", 15, 16 ),
		new Level( "d w s s sc cs cy f f", 23 ),
		new Level( "d d w w pg wt wt f f", 26, 27 )
	}, 1 )
	};
	/* some Chaos Mode levels
	d d sc dn dn cs dy wt cp b
	sc sc sc w dn dn cs sf sh b
	d d w dn cs cy sh b b ps
	*/
	// tutorial, basic, intermediate, tricky, complex, advanced, expert, stupefying, mind-bending, genius
}

File: TrickyTropolis/Assets/Code/Old Java/LevelGroup.java
package sequence;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Scanner;

class Level {
	public GameCard[] startHand;
	public int energyGoal;
	public String title;
	public String description;
	Level( GameCard[] h, int e, String t, String d ) {
		startHand = h; energyGoal = e; title = t; description = d;
	}
}

public class LevelGroup {
	private LinkedList<Level> list = new LinkedList<Level>();
	public Level[] list() { return list.toArray(new Level[list.size()]); }
	public LevelGroup add( String handCode, int e, String t, String d ) {
		LinkedList<GameCard> cards = new LinkedList<GameCard>();
		Scanner s = new Scanner(handCode);
		while( s.hasNext() ) cards.add( code.get(s.next()) );
		s.close();
		list.add( new Level (cards.toArray(new GameCard[0]),e,t,d) );
		return this;
	}
	
	private static final Map<String, GameCard> code = new HashMap<String, GameCard>();
	public static int NUM_GROUPS = 3;
	public static LevelGroup[] levelGroups = new LevelGroup[NUM_GROUPS];
	public static void loadLevels() {
		for( GameCard c : BuildingCard.list() ) code.put( c.code, c );
		for( GameCard c : EventCard.list() ) code.put( c.code, c );
		
		/////////////////////////////////////////
		//Tutorial levels
		/////////////////////////////////////////
		
		LevelGroup tutorialLevels = new LevelGroup()
		.add( "w w w w", 8,
			"Welcome to Sequence!", 
			"The object of this game is to reach the total Energy goal for each level.\n"
			+ "Do this by playing buildings from your hand (left), each of which holds a \n"
			+ "certain amount of Energy. The counter above the playing field shows how \n"
			+ "close you are to your goal.\n\n"
			+ "For this level, simply play all four Warehouses to win.  You must place \n"
			+ "them in different slots or you will have to restart (spacebar)." )
			
		.add( "w w cs cs", 8, 
			"Energy Requirement",
			"Pay attention to the numbers in the upper right corner of your cards!\n"
			+ "The first number is Energy Cost; the second number is Starting Energy.\n\n"
			+ "Because Charging Station costs 2, it cannot be played on an empty square.\n"  
			+ "Instead, it must replace an existing building having at least 2\n"
			+ "Energy. That building and its Energy will be lost, so choose carefully!" ) 
			
		.add( "w w w w s", 9, 
			"Conduit",
			"Buildings with Conduit gain +1 Energy whenever another building\n"
			+ "is played.  The sooner you get Conduits out, the more Energy\n"
			+ "they will gain.")
			
		.add( "d d cs", 5, 
			"UPGRADE!",
			"The UPGRADE attribute gives a bonus to whichever building\n"
			+ "is constructed over it.  When you play a building over a \n"
			+ "Databank, the new building gets 1 more Energy than normal." )
			
		.add( "d w cs cs", 8, 
			"Energy Precision",
			"In genereral, you get the most value out of your buildings when you meet \n"
			 + "their Energy requirements exactly. Avoid waste!")
			 
		.add( "w w d wt", 6, 
			"Supply",
			"When you have a building with Supply in play, it will provide \n"
			+ "+1 Energy to every new building you make. Like Conduit, this \n"
			+ "ability gives more Energy the earlier you can get it.")
			
		.add("w d d cs pu", 10, 
			"Event Cards",
			"Event cards, such as Power-Up, give Energy or other bonuses to your \n"
			+ "existing buildings. Figuring out the best time to play them can be tricky!" )
			
		.add("d i i s", 5, 
			"Industrial Zone",
			"The Industrial Zone creates two buildings with no starting Energy.\n"
			+ "This is useful for triggering effects like Conduit and Supply." )
			
		.add( "w s cs cs cy", 6,
			"Construction",
			"The Construction ability reduces the Energy Cost of the buildings\n"
			+ "in your hand by 1. If a building's cost is reduced to zero,\n"
			+ "you can play in on an empty slot!\n\n"
			+ "Hint: You can replace the Construction Yard itself and still\n"
			+ "get the cost reduction!" )

		.add("d d cs rf ps", 8,
			"Energy Multiplier",
			"The Research Facility gains double Energy from all sources!\n"
			+ "This makes it a great target for Energy Boost, Supply, and so on.");
		
		/////////////////////////////////////////
		//Intermediate levels
		/////////////////////////////////////////
		
		LevelGroup intermediateLevels = new LevelGroup()
		.add( "d d i s s pu", 12,
			"Intermediate Levels", 
			"Now that you know the basics, let's start to make things a \n"
			+"little harder. You'll start with more cards and fewer hints." )
			
		.add( "i i i s s pu pu", 15,
			"Zoning Code", 
			"Get ready for some serious Industrial Zoning!" )
			
		.add( "w w w s s s s s", 7,
				"Too many Substations",
				"No hints this time..." )
			
		.add( "i w s cs pu pu",  11,
			"Precision vs Sequencing",
			"Sometimes you will have to choose between meeting an Energy \n"
			+"requirement exactly, versus playing cards in the more logical order.\n"
			+"Always be aware of your choices!" )
				
		.add( "d w s cy wt ps", 8,
			"Ability Choices", 
			"Conduit, Supply, and Construction all want to be played \n"
			+"early. So which comes first?  That depends, of course!" )
		
		.add( "d i w sc sc wt rf", 11,
			"Storage Cells",
			"Storage cells can be used to concentrate Energy in one place."
			+"They have other uses too!")
			
		.add( "d d i w sc wt", 10,
			"Storage Cells 2",
			"Can you get that last point of energy?")
			
		.add( "i d w s wt cy pu", 11,
				"Red Herring",
				"Occasionally you'll be given cards which don't actually need.\n"
				+"Don't be fooled!  Playing all your cards is not a requirement\n"
				+"for victory!")
		
		.add( "d w s cs rf ps c", 13, 
				"Science!",
				"Time to fund some serious research.")
				
		.add( "d w w s wt wt wt wt", 9,
				"A Quixotic Level",
				"Try to make the most of a weird hand." );
		
		/////////////////////////////////////////
		//Advanced levels
		/////////////////////////////////////////
		LevelGroup advancedLevels = new LevelGroup()
		.add( "i sc w w cs wt c sk sk", 26,
			"Advanced Levels", 
			"Welcome to the Advanced levels!  There will be some bigger cards and \n"+ 
			"even bigger hands!")
							
		.add( "d d i i i w cs cp ps", 16,
			"Nonrenewable Resource",
			"The Coal plant gives 2 Supply, but loses 1 Energy with each play. If\n" +
			"it runs out of Energy, it self-destructs!")
			
		.add( "d w w i cs wt wt wt ps cp", 25, 
			"Supply Center",
			"Can you manage your resources effectively?" )
		
		.add( "d w s s sc cs cy f f", 23, 
			"Assembly Line",
			"Don't be afraid to use trial and error, but it has to be guided by a sense\n"+
			"of what the possible good plays are." )
			
		.add( "d d w w rc wt wt f f", 26,
			"Reactor Core",
			"The Reactor Core gives an extra Supply to any building constructed over it.\n"+
			"Bonus: 27 Energy is possible!" )

		.add( "d d i w rc cs cs wt rf", 24,
			"Reactor Core II",
			"Can you spot the clever opening?" )
		
		.add( "w w cs ms ms ms f sk sk", 24,
			"Machine Shop", 
			"The Machine Shop is a flexible building which can upgrade itself or another\n"+
			"building with Conduit." )
			
		.add( "d d w tl cs cy ms sk ps", 23, 
			"Tech Lab",
			"The Tech Lab rapidly gains Energy when you play Event cards." )
			
		.add( "d w w s rc cs cy cp f rf sk", 26,
			"Heavy-handed",
			"You have many high-cost buildings.  Can you make use of them all?\n"+
			"28 Energy is possible!" )
			
		.add( "d d w i s cs wt ms cp rf", 32,
			"Science II",
			"Last level (for now)!  Don't give up!"  );
				
		levelGroups[0] = tutorialLevels; 
		levelGroups[1] = intermediateLevels;
		levelGroups[2] = advancedLevels;
	}	
}

	
	
	
	
	

